tasks/change-granule-collection-s3/src/index.ts

```diff
@@ -0,0 +1,388 @@
+  }
+  const [match] = matches;
+  if (!bucketsConfig.keyExists(match.bucket)) {
+    throw new InvalidArgument(`Collection config specifies a bucket key of ${match.bucket}, `
+      + `but the configured bucket keys are: ${Object.keys(bucketsConfig).join(', ')}`);
+  }
+  return match;
+}
+
+/**
+ * Copy granule files in s3.
+ * Any CMRfile will not truly be "copied" but pushed up to new location with new metadata contents
+ * any other file will be checked to be sure a copy is actually needed, and then copied over
+ */
+async function copyGranulesInS3({
+  sourceGranules,
+  targetGranules,
+  cmrObjects,
+  s3MultipartChunksizeMb,
+}: {
+  sourceGranules: Array<ValidGranuleRecord>,
+  targetGranules: Array<ValidGranuleRecord>,
+  cmrObjects: { [key: string]: Object },
+  s3MultipartChunksizeMb?: number,
+}): Promise<void> {
+  await pMap(
+    zip(sourceGranules, targetGranules),
+    async ([sourceGranule, targetGranule]) => {
+      if (!sourceGranule?.files || !targetGranule?.files) {
+        throw new AssertionError({ message: 'size mismatch between target and source granules' });
+      }
+      return Promise.all(zip(sourceGranule.files, targetGranule.files)
+        .map(async ([sourceFile, targetFile]) => {
+          if (!(sourceFile && targetFile)) {
+            throw new AssertionError({
+              message: 'size mismatch between target and source granule files',
+            });
+          }
+          const isMetadataFile = isCMRMetadataFile(targetFile);
+          if (isMetadataFile) {
+            await uploadCMRFile(targetFile, cmrObjects[targetGranule.granuleId]);
+            return;
+          }
+          if (!await s3MoveNeeded(sourceFile, targetFile)) {
+            return;
+          }
+          if (!isMetadataFile) {
+            await copyObject({
+              sourceBucket: sourceFile.bucket,
+              sourceKey: sourceFile.key,
+              destinationBucket: targetFile.bucket,
+              destinationKey: targetFile.key,
+              chunkSize: s3MultipartChunksizeMb,
+            });
+          }
+        }));
+    },
+    { concurrency: getConcurrency() }
+  );
+}
+
+/**
+ * Create new file object updated to new collection data
+ */
+function updateFileMetadata(
+  file: Omit<ValidApiFile, 'granuleId'>,
+  granule: ValidGranuleRecord,
+  bucketsConfig: BucketsConfig,
+  cmrMetadata: Object,
+  targetCollection: CollectionRecord
+): Omit<ValidApiFile, 'granuleId'> {
+  const fileName = path.basename(file.key);
+  const match = identifyFileMatch(bucketsConfig, fileName, targetCollection.files);
+  const URLPathTemplate = match.url_path || targetCollection.url_path || '';
+  const urlPath = urlPathTemplate(URLPathTemplate, {
+    file,
+    granule: granule,
+    cmrMetadata,
+  });
+  const updatedBucket = bucketsConfig.nameByKey(match.bucket);
+  const updatedKey = S3.s3Join(urlPath, fileName);
+  return {
+    ...cloneDeep(file),
```

Comment:
ApiFile doesn't have any nested objects, use of the spread operator here is probably acceptable and more efficient.